cert/aziot-certd/src/lib.rs

```diff
@@ -195,654 +258,349 @@ impl UpdateConfig for Api {
-                                        })?;
-
-                                        // EST identity cert was obtained and persisted successfully. Now recurse to retry the original cert request.
-
-                                        let x509 = create_cert(api, id, csr, issuer).await?;
-                                        Ok(x509)
-                                    }
-
-                                    Err(bootstrap_identity_err) => {
-                                        // Neither EST identity cert nor EST bootstrap identity cert could be obtained.
-                                        Err(Error::Internal(InternalError::CreateCert(format!(
-                                            "cert {:?} is configured to be issued by EST, but neither EST identity nor EST bootstrap identity could be obtained; \
-                                            {} {}",
-                                            id,
-                                            identity_err,
-                                            bootstrap_identity_err,
-                                        ).into())))
-                                    }
-                                }
-                            }
+                            Ok((cert, pk))
                         }
-                    } else {
-                        // We need to only use basic auth with the EST server.
-
-                        let x509 = est::create_cert(
-                            csr,
-                            url,
-                            auth_basic,
-                            None,
-                            trusted_certs_x509,
-                            api.proxy_uri.clone(),
-                        )
-                        .await?;
-
-                        let path = aziot_certd_config::util::get_path(
-                            &api.homedir_path,
-                            &api.preloaded_certs,
-                            id,
-                            true,
-                        )
-                        .map_err(|err| Error::Internal(InternalError::GetPath(err)))?;
-                        std::fs::write(path, &x509).map_err(|err| {
-                            Error::Internal(InternalError::CreateCert(Box::new(err)))
-                        })?;
+                    })
+                    .transpose();
+
+                match id_opt {
+                    Ok(id_opt) =>
+                        est::create_cert(
+                                base64::encode(req.to_der()?).into_bytes(),
+                                url,
+                                auth.basic.as_ref(),
+                                id_opt.as_ref().map(|(cert, pk)| (&**cert, &**pk)),
+                                &trusted_certs,
+                                api.proxy_uri.clone()
+                            )
+                            .await,
+                    Err(err) => {
+                        log::warn!("{}", err);
```

Comment:
This should be bubbling up the error like the old code did so that the context chain is preserved, rather than ad-hoc logging like this.

Reply:
Addendum: the current code does not fall back to the bootstrap certificate when fetching with the identity fails, so the only kinds of errors that will be propagated will be identity fetching errors. However, this is the same behavior as the original code. Should we fall back to the bootstrap certificate when the identity certificate fails to produce a valid certificate (may be incorrect if bootstrap is single-use) and/or try re-enrolling?